rbs/prism/AssertionsRewriterPrism.cc

```diff
@@ -1,16 +1,1140 @@
+        //   - Pair[X] where X is a type parameter -> T.type_parameter(:X)
+        //   - The :X symbol needs to be copied with proper string allocation
+        case PM_SYMBOL_NODE: {
+            auto *original = down_cast<pm_symbol_node_t>(node);
+            auto symbolStr = parser.extractString(&original->unescaped);
+            auto loc = parser.translateLocation(original->base.location);
+            return prism.Symbol(loc, symbolStr);
+        }
+
+        default:
+            unreachable("deepCopyGenericTypeNode: unhandled node type {} - this node type appears in generic type "
+                        "instantiations but is not yet supported",
+                        PM_NODE_TYPE(node));
+    }
+}
+
+/**
+ * If `node` is a `.new` call and `type` is a generic type instantiation of the same class,
+ * replace the receiver of `.new` with the generic type.
+ *
+ * For example: `G1.new #: G1[Integer]` becomes `G1[Integer].new`.
+ */
+void maybeSupplyGenericTypeArguments(core::MutableContext ctx, parser::Prism::Parser &parser, pm_node_t **node,
+                                     pm_node_t **type) {
+    // We only rewrite `.new` calls
+    if (!PM_NODE_TYPE_P(*node, PM_CALL_NODE)) {
+        return;
+    }
+
+    auto *newCall = down_cast<pm_call_node_t>(*node);
+    auto methodName = parser.resolveConstant(newCall->name);
+
+    if (methodName != "new"sv) {
+        return;
+    }
+
+    // We only rewrite when casted to a generic type (syntheticSquareBrackets)
+    if (!PM_NODE_TYPE_P(*type, PM_CALL_NODE)) {
+        return;
+    }
+
+    auto *bracketCall = down_cast<pm_call_node_t>(*type);
+    auto bracketMethodName = parser.resolveConstant(bracketCall->name);
+
+    if (bracketMethodName != "<syntheticSquareBrackets>"sv) {
+        return;
+    }
+
+    // We only rewrite when the generic type is the same as the instantiated one
+    if (!sameConstant(parser, newCall->receiver, bracketCall->receiver)) {
+        return;
+    }
+
+    // We need to create a deep copy of the type node because it will be used in two places:
+    // 1. As the receiver of .new()
+    // 2. As the type argument to T.let()
+    newCall->receiver = deepCopyGenericTypeNode(parser, *type);
+}
+
+} // namespace
+
+/**
+ * Save the signature from the given block so it can be used to resolve the type parameters from the method signature.
+ *
+ * Returns `true` if the block is a `sig` call, `false` otherwise.
+ */
+bool AssertionsRewriterPrism::saveTypeParams(pm_node_t *call) {
+    if (!call || !PM_NODE_TYPE_P(call, PM_CALL_NODE)) {
+        return false;
+    }
+
+    auto *callNode = down_cast<pm_call_node_t>(call);
+
+    // Check if this is a sig() call
+    auto methodName = parser.resolveConstant(callNode->name);
+    if (methodName != "sig") {
+        return false;
+    }
+
+    // Check if it has a block
+    if (!callNode->block) {
+        return false;
+    }
+
+    typeParams = extractTypeParamsPrism(ctx, parser, callNode->block);
```

Comment:
In this case, I think it'd be worth using `this->` to clarify that we're mutating the context of the `AssertionsRewriterPrism` itself. (Pun intended)


```suggestion
    this->typeParams = extractTypeParamsPrism(ctx, parser, callNode->block);
```